packages/babel-core/src/config/files/module-types.ts

```diff
@@ -23,38 +26,96 @@ export const supportsESM = semver.satisfies(
   "^12.17 || >=13.2",
 );
 
-export default function* loadCjsOrMjsDefault(
+export default function* loadCodeDefault(
   filepath: string,
   asyncError: string,
   // TODO(Babel 8): Remove this
   fallbackToTranspiledModule: boolean = false,
 ): Handler<unknown> {
-  switch (guessJSModuleType(filepath)) {
-    case "cjs":
+  switch (path.extname(filepath)) {
+    case ".cjs":
       return loadCjsDefault(filepath, fallbackToTranspiledModule);
-    case "unknown":
+    case ".mjs":
+      break;
+    case ".cts":
+      return loadCtsDefault(filepath);
+    default:
       try {
         return loadCjsDefault(filepath, fallbackToTranspiledModule);
       } catch (e) {
         if (e.code !== "ERR_REQUIRE_ESM") throw e;
       }
-    // fall through
-    case "mjs":
-      if (yield* isAsync()) {
-        return yield* waitFor(loadMjsDefault(filepath));
-      }
-      throw new ConfigError(asyncError, filepath);
   }
+  if (yield* isAsync()) {
+    return yield* waitFor(loadMjsDefault(filepath));
+  }
+  throw new ConfigError(asyncError, filepath);
 }
 
-function guessJSModuleType(filename: string): "cjs" | "mjs" | "unknown" {
-  switch (path.extname(filename)) {
-    case ".cjs":
-      return "cjs";
-    case ".mjs":
-      return "mjs";
-    default:
-      return "unknown";
+function loadCtsDefault(filepath: string) {
+  const ext = ".cts";
+  const hasTsSupport = !!(
+    require.extensions[".ts"] ||
+    require.extensions[".cts"] ||
+    require.extensions[".mts"]
+  );
+
+  let handler: NodeJS.RequireExtensions[""];
+
+  if (!hasTsSupport) {
+    const preset = require("@babel/preset-typescript");
+
+    if (!preset) {
+      throw new ConfigError(
+        "You appear to be using a .cts file as Babel configuration, but the `@babel/preset-typescript` package was not found, please install it!",
+        filepath,
+      );
+    }
+
+    const opts: InputOptions = {
+      babelrc: false,
+      configFile: false,
+      sourceType: "script",
+      sourceMaps: "inline",
+      presets: [
+        [
+          preset,
+          {
+            disallowAmbiguousJSXLike: true,
+            allExtensions: true,
+            onlyRemoveTypeImports: true,
+            optimizeConstEnums: true,
+            ...(process.env.BABEL_8_BREAKING && {
+              allowDeclareFields: true,
+            }),
```

Comment:
```suggestion
            ...(!process.env.BABEL_8_BREAKING && {
              allowDeclareFields: true,
            }),
```
In Babel 8 this field is removed, right?

Reply:
Oops, thanks!
I'm surprised CI didn't fail.

Reply:
I get it, it's only not allowed to be set in babel8.
Glad the new plugin logic is bug free. :)